### DISCUSSION TITLE
Typesafe way to re map tuple

### DISCUSSION THREAD

**Discussion by b3nb5n** [colinhacks/zod#1811] [Q&A] [answered]
I would like to write a function that can transform the elements of a zod tuple in a typesafe way. the goal is to make all elements in a tuple optional i.e `z.tuple([z.string(), z.number()])` => `z.tuple([z.string().optional(), z.number().optional()])`. so far I've written `z.tuple(inputSchema.items.map((schema) => schema.optional()))` which seems to work at runtime, but causes typescript errors and doesn't preserve the types/order of the tuple items when using `z.infer`. any ideas?

**Comment by JacobWeisenburger** [comment, marked as answer]
I'm not sure this is possible due to limitations in typescript.

**Comment by JacobWeisenburger** [comment]
Making all properties optional is really easy if you use an object instead of a tuple:

```ts
const schema = z.object( {
    string: z.string(),
    number: z.number(),
} )
type Schema = z.infer<typeof schema>
// type Schema = {
//     string: string
//     number: number
// }
console.log( schema.safeParse( {} ).success ) // false

const schema2 = schema.partial()
type Schema2 = z.infer<typeof schema2>
// type Schema2 = {
//     string?: string | undefined
//     number?: number | undefined
// }
console.log( schema2.safeParse( {} ).success ) // true
```
